packages/engine/Source/Renderer/ShaderSource.js

```diff
@@ -278,19 +266,38 @@ function combineShader(shaderSource, isFragmentShader, context) {
   }
 
   // append built-ins
+  let builtinSources = "";
   if (shaderSource.includeBuiltIns) {
-    result += getBuiltinsAndAutomaticUniforms(combinedSources);
+    builtinSources = getBuiltinsAndAutomaticUniforms(combinedSources);
   }
 
   // reset line number
   result += "\n#line 0\n";
 
   // append actual source
+  const combinedShader = builtinSources + combinedSources;
+  if (
+    context.webgl2 &&
+    isFragmentShader &&
+    !/layout\s*\(location\s*=\s*0\)\s*out\s+vec4\s+out_FragColor;/g.test(
+      combinedShader
+    ) &&
+    !/czm_out_FragColor/g.test(combinedShader) &&
+    /out_FragColor/g.test(combinedShader)
+  ) {
+    result += "layout(location = 0) out vec4 out_FragColor;\n\n";
+  }
+
+  result += builtinSources;
   result += combinedSources;
 
   // modernize the source
-  if (context.webgl2) {
-    result = modernizeShader(result, isFragmentShader, true);
+  if (!context.webgl2) {
+    result = demodernizeShader(result, isFragmentShader);
+  } else {
+    // Replace all czm_textureCube calls with texture
+    result = result.replaceAll(/czm_textureCube/g, `texture`);
+    result = `#version 300 es\n${result}`;
```

Comment:
Should `czm_textureCube` be an actual file `textureCube.glsl` that looks like:

```glsl
vec4 czm_textureCube(in samplerCube sampler, in vec3 p) {
#if __VERSION__ == 300
    return texture(sampler, p);
#else  
    return textureCube(sampler, p);
#endif
}
```

Reply:
Good idea! I also didn't know you could do the version check like that. It's handy!